File: server/Models/FocusSession.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace TasksTrack.Models
{
    public class FocusSession
    {
        public int Id { get; set; }

        [Required(ErrorMessage = "HabitId is required")]
        public int HabitId { get; set; }

        [Required(ErrorMessage = "CreatedBy is required")]
        public required string CreatedBy { get; set; }

        [Required(ErrorMessage = "StartTime is required")]
        public DateTimeOffset StartTime { get; set; }

        public DateTimeOffset? PauseTime { get; set; }
        public DateTimeOffset? ResumeTime { get; set; }
        public DateTimeOffset? EndTime { get; set; }

        [Required(ErrorMessage = "Status is required")]
        [StringLength(FocusSessionConstants.MAX_STATUS_LENGTH, ErrorMessage = "Status cannot exceed {1} characters")]
        public required string Status { get; set; } // "active", "paused", "completed", "interrupted"

        public int PlannedDurationMinutes { get; set; } = FocusSessionConstants.DEFAULT_POMODORO_MINUTES; // Default Pomodoro time
        public int? ActualDurationSeconds { get; set; } // Actual time worked
        public int? PausedDurationSeconds { get; set; } // Total time paused

        [StringLength(FocusSessionConstants.MAX_NOTES_LENGTH, ErrorMessage = "Notes cannot exceed {1} characters")]
        public string? Notes { get; set; }

        public DateTimeOffset CreatedDate { get; set; }
        public DateTimeOffset? UpdatedDate { get; set; }
        public string? UpdatedBy { get; set; }

        // Navigation properties
        [ForeignKey("HabitId")]
        public Habit? Habit { get; set; }
    }
}

File: client/src/types/focusSession.ts
import { z } from 'zod';
import { habitEntitySchema } from './habit';

// Focus session status enum schema
export const focusSessionStatusSchema = z.enum(['active', 'paused', 'completed', 'interrupted']);

// Focus session entity schema - matches FocusSessionResponse from backend
export const focusSessionEntitySchema = z.object({
  id: z.number(),
  habitId: z.number(),
  startTime: z.string(), // DateTimeOffset serialized as string
  endTime: z.string().optional(), // DateTimeOffset? serialized as string or null
  plannedDurationMinutes: z.number(),
  notes: z.string().optional(),
  createdBy: z.string().optional(), // nullable in response
  pauseTime: z.string().optional(), // DateTimeOffset? serialized as string or null
  resumeTime: z.string().optional(), // DateTimeOffset? serialized as string or null
  status: z.string(), // Server sends as string
  actualDurationSeconds: z.number(), // Non-nullable in response (defaults to 0)
  pausedDurationSeconds: z.number(), // Non-nullable in response (defaults to 0)
  createdDate: z.string(), // DateTimeOffset serialized as string
  habit: habitEntitySchema.optional(), // Nested Habit object (nullable)
});

// Focus session create schema
export const focusSessionCreateSchema = focusSessionEntitySchema.pick({
  habitId: true,
  plannedDurationMinutes: true,
  notes: true,
}).partial({
  plannedDurationMinutes: true,
  notes: true,
});

// Focus session update schema
export const focusSessionUpdateSchema = focusSessionEntitySchema.pick({
  endTime: true,
  notes: true,
  status: true,
  actualDurationSeconds: true,
  pausedDurationSeconds: true,
}).partial();

// Focus session analytics schema - matches FocusSessionAnalytics from backend
export const focusSessionAnalyticsSchema = z.object({
  totalSessions: z.number(),
  totalMinutes: z.number(),
  completedSessions: z.number(),
  averageSessionMinutes: z.number(),
  longestSessionMinutes: z.number(),
  currentStreak: z.number(),
  longestStreak: z.number(),
  completionRate: z.number(),
});

// Export types
export type IFocusSession = z.infer<typeof focusSessionEntitySchema>;
export type IFocusSessionCreateRequest = z.infer<typeof focusSessionCreateSchema>;
export type IFocusSessionUpdateRequest = z.infer<typeof focusSessionUpdateSchema>;
export type IFocusSessionAnalytics = z.infer<typeof focusSessionAnalyticsSchema>;
export type FocusSessionStatus = z.infer<typeof focusSessionStatusSchema>;

// Export enum for use in switch statements and value comparisons
export const FocusSessionStatus = {
  Active: 'active',
  Paused: 'paused',
  Completed: 'completed',
  Interrupted: 'interrupted'
} as const;
